### PR TITLE
Fix type hints and robustify key model attributes/properties

### PR DESCRIPTION
**Pull request by michaelosthege**
This fixes all type problems detected by `mypy`, and ensures consistency moving forward by running it in the pre-commit.

Version was bumped already in #63.

### CODE REVIEW

File: calibr8/test_all.py [hunk 1]
```diff
@@ -187,9 +187,37 @@ def test_parametrization_equivalence(self, cls: calibr8.DistributionMixin, a, b,
...
+                x=numpy.arange(3),
+                y=[1, 2, 3],
+                theta=theta,
+            )
+        pass
+
     def test_no_logpdmf_error(self):
         class InvalidNoise(calibr8.DistributionMixin):
-            scipy_dist = None
+            scipy_dist = object()
```
Comment (michaelosthege) [inline calibr8/test_all.py:220]: because of the new `None` check prior to the exception to be tested

File: calibr8/utils.py [hunk 2]
```diff
@@ -43,8 +44,10 @@ def __getattr__(self, attr):
 try:
     import pymc as pm
 
+    DistributionType: TypeAlias = Optional[pm.Distribution]
     HAS_PYMC = True
 except ModuleNotFoundError:
+    DistributionType: TypeAlias = Literal[None]  # type: ignore
```
Comment (michaelosthege) [inline calibr8/utils.py:50]: mypy doesn't like types to be redefined, but this is the only way I'm aware of to do type hinting with optional dependencies

File: calibr8/utils.py [hunk 3]
```diff
@@ -100,10 +103,9 @@ def istensor(input: object) -> bool:
             if istensor(element):
                 return True
     elif isinstance(input, Iterable):
-        if len(input) > 1:
-            for element in input:
-                if istensor(element):
-                    return True
+        for element in input:
+            if istensor(element):
+                return True
```
Comment (michaelosthege) [inline calibr8/utils.py:108]: because `Iterable` is not necessarily `Sized`

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -19,3 +19,8 @@ repos:
   rev: 22.3.0
   hooks:
     - id: black
+- repo: https://github.com/pre-commit/mirrors-mypy
+  rev: v1.3.0
+  hooks:
+  -   id: mypy
+      exclude: 'test_.*?\.py$'
```
Comment (lhelleckes) [inline .pre-commit-config.yaml:26]: What does it do here to ignore the test files?

File: calibr8/utils.py (hunk 2 above)
Comment (lhelleckes) [reply calibr8/utils.py:50]: Okay, give me a little help here. The TypeAlias is actually defining attributes here?

File: .pre-commit-config.yaml (hunk 4 above)
Comment (michaelosthege) [reply .pre-commit-config.yaml:26]: Yes, ignores the test files which (in this project) sit together with the others.
Some tests use incorrect types on purpose..

File: calibr8/utils.py (hunk 2 above)
Comment (michaelosthege) [reply calibr8/utils.py:50]: `TypeAlias` is just the type hint to let mypy know that `DistributionType` is an alias for a type - in this case `Literal[None]`

does that answer your question?

**Review by lhelleckes — APPROVED** [review]
Thx, much nicer.

**Pull request merged**